### PR TITLE
[wip] comparing vanilla torch model and clipping with OLMo FSDP, no_shard and OLMo clipping

### PR DESCRIPTION
**Pull request by ananyahjha93**
image

There is an order of magnitude difference between the losses between the two setups. @dirkgr @epwalsh can you sanity check the OLMo grad_clipping code for FSDP no_shard/DDP?

On 3 batches being sent to the model again and again, the model should overfit and the loss should go to 0. We can see that in the screenshot for vanilla pytorch run.

Comparing the two runs:
1. OLMo model used in both
2. no FSDP vs FSDP-no_shard (32 bit)
3. torch optim and clipping vs OLMo optim and clipping
4. OLMo scheduler used in both

### CODE REVIEW

**Comment by dirkgr** [comment]
Can you make this a draft PR? I don't think we'll merge this?

**Comment by dirkgr** [comment]
* no FSDP vs FSDP-no_shard (32 bit)
 * torch optim and clipping vs OLMo optim and clipping
 
 Find out which one of these makes the difference?

File: tests/grad_norm_test.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    # build fsdp config
+    fsdp_model = FSDP(
+        model,
+        sharding_strategy=ShardingStrategy.NO_SHARD,
+        mixed_precision=MixedPrecision(
+            param_dtype=torch.float32,
+            reduce_dtype=torch.float32,
+            buffer_dtype=torch.float32,
+        ),
+        auto_wrap_policy=None,
```
Comment (dirkgr) [inline tests/grad_norm_test.py:217]: Did you check how it wraps the model?

File: tests/grad_norm_test.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+    del model
+    del torch_optimizer
+    del scheduler
+    del data_loader
+
+    # use same model, data, optimizer, fsdp_model and send to trainer and compare gradient clip
+
+    # olmo optimizer
+    model = OLMo(cfg.model).to('cuda')
```
Comment (dirkgr) [inline tests/grad_norm_test.py:194]: Can you make double sure the initialization is same here?

File: tests/grad_norm_test.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    olmo_optimizer = build_optimizer(cfg, model)
+    data_loader = build_train_dataloader(cfg)
```
Comment (dirkgr) [inline tests/grad_norm_test.py:196]: Make sure the data loader is same too? By the time you get here you might have consumed some random state, so you might get different data. I would go so far as to pre-load the data into a `List`, and then use that, so you can be 100% sure it's the same.

File: tests/grad_norm_test.py (hunk 2 above)
Comment (epwalsh) [reply tests/grad_norm_test.py:194]: From the looks of it, this model will definitely start with different weights than the one above. The best way to be sure would be to load starting state dict of the first model into the second model.

File: olmo/optim.py [hunk 4]
```diff
@@ -141,10 +140,17 @@ def clip_grads_and_collect_metrics(
...
+        #######################################################################
+        param_group_sharded = False
+        for group in self.param_groups:
+            param_group_sharded = param_group_sharded or group.get("sharded", False)
+
         # Now reduce metrics over all ranks.
         total_grad_norm: torch.Tensor
         per_param_avg_metrics: List[torch.Tensor] = []
-        if is_distributed():  # TODO (epwalsh): skip for non-sharded params
+        if is_distributed() and param_group_sharded:
```
Comment (epwalsh) [inline olmo/optim.py:153]: Will still need to reduce gradient metrics with non-sharded params since each rank will have different gradients.

Comment (ananyahjha93) [reply olmo/optim.py:153]: won't the gradients sync after loss.backward() call?

Comment (epwalsh) [reply olmo/optim.py:153]: Ah, yes you're right. My mistake

Comment (ananyahjha93) [reply olmo/optim.py:153]: in DDP loss.backward() syncs grads, optimizer.step() updates each copy with synced grads! Or am I getting this wrong? So, if loss.backward() syncs grads, then every rank must have the same grads

**Comment by ananyahjha93** [comment]
image

Sending torch optimizer and gradient clipping with no_shard seems to fix it!

File: olmo/optim.py (hunk 4 above)
Comment (ananyahjha93) [reply olmo/optim.py:153]: @epwalsh does this apply to FSDP no_shard as well?
